### DISCUSSION TITLE
What's the idiomatic way to use external module in macro?

### DISCUSSION THREAD

**Discussion by tupini07** [janet-lang/janet#988] [Q&A] [answered]
As the title says, what's considered the most idiomatic way to use a module inside a macro. For example, the naive (dirty) way would be to 

```clojure
# macros.janet
(import some-module)

(defmacro a-macro []
    ~(some-module/function params))
```

and then use it as

```clojure
# some-file.janet
(import some-module) # < need to import this even if we don't use it directly
(import ./macros)

(a-macro)
```

The issue here is that `some-file.janet` needs to import `some-module`, even though it's not using it directly. Is there a way in which I can require the module directly inside the body of the macro so I don't need users to import it separately? If `import` gave me a table of symbols I imagine this could be done with `let` as:

```clojure
(let [sm (import some-module)]
   (sm/function params))
```

But unless I missed something while going through the documentation there is no way to do this in Janet 🤔

**Comment by bakpakin** [comment, marked as answer]
The `as-macro` macro is a hack for this.

```
(import my-other-module)

(defmacro my-second-macro
  [x y z]
  ~(+ ,x
      (,my-other-module/some-fn ,y)
      (as-macro ,my-other-module/first-macro ,z)))
```

It allows you to use the same unquote trick for hygiene on macros as you use on functions. Note that as-macro is not unquoted.

**Reply by bakpakin** [reply]
This also has the benefit that you don't need to import `my-other-module` when using `my-second-macro` - the macro is essentially inlined for you. The caveat is that you cannot override the `as-macro` binding or this breaks, but you shouldn't be overriding core bindings anyway in most code.

**Reply by tupini07** [reply]
Great, thanks a lot! That's exactly what I was looking for. I probably missed it while going through the documentation on Janet's site
